internal/config/inputs.go

```diff
@@ -0,0 +1,146 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/sethvargo/go-envconfig"
+)
+
+// Inputs roughly maps to the set of action inputs.
+type Inputs struct {
+	// ProductRepository isn't really an input, but it's
+	// required for creating default values.
+	ProductRepository string `env:"PRODUCT_REPOSITORY"`
+
+	// Required inputs.
+	ProductName    string `env:"PRODUCT_NAME"`
+	ProductVersion string `env:"PRODUCT_VERSION"`
+	GoVersion      string `env:"GO_VERSION"`
+	OS             string `env:"OS"`
+	Arch           string `env:"ARCH"`
+	Reproducible   string `env:"REPRODUCIBLE"`
+	Instructions   string `env:"INSTRUCTIONS"`
+
+	// Optional inputs.
+	BinName string `env:"BIN_NAME"`
+	ZipName string `env:"ZIP_NAME"`
+
+	// Inputs only used in testing.
+	PrimaryBuildRoot      string `env:"PRIMARY_BUILD_ROOT"`
+	VerificationBuildRoot string `env:"VERIFICATION_BUILD_ROOT"`
+}
+
+type dirNames struct {
+	target, zip, meta string
+}
+
+var dirs = dirNames{"dist", "out", "meta"}
+
+func (i Inputs) Config(rc RepoContext) (Config, error) {
+	i = i.trimSpace().setDefaults(rc)
+	if err := i.validate(); err != nil {
+		return Config{}, err
+	}
+	return Config{
+		Inputs:              i,
+		ProductRevision:     rc.CommitSHA,
+		ProductRevisionTime: rc.CommitTime.UTC().Format(time.RFC3339),
+		PrimaryBuild:        newBuildConfig(dirs, i.PrimaryBuildRoot, i.BinName, i.ZipName),
+		VerificationBuild:   newBuildConfig(dirs, i.VerificationBuildRoot, i.BinName, i.ZipName),
+		TargetDir:           dirs.target,
+		ZipDir:              dirs.zip,
+		MetaDir:             dirs.meta,
+	}, nil
+}
+
+// FromEnvironment creates a new Config from environment variables.
+// See Inputs for a list of the environment variables read by Digest.
+func FromEnvironment() (Config, error) {
+	ctx := context.Background()
+
+	rc, err := readRepoContext()
+	if err != nil {
+		return Config{}, err
+	}
+
+	var inputs Inputs
+	if err := envconfig.Process(ctx, &inputs); err != nil {
+		return Config{}, err
+	}
+
+	return inputs.Config(rc)
+}
+
+// trimSpace trims leading and trailing whitespace from every input string.
+func (i Inputs) trimSpace() Inputs {
```

Comment:
Is there any particular reason `i` isn't a pointer? Not arguing either way, just curious. It seems like in this function and the ones following it could be a reference to an `Inputs` which modifies itself

Reply:
I usually have a preference for writing Go in a functional style, i.e. without mutating functions, when possible, because it makes the code easier to follow. I.e. you can call functions and the caller gets to decide what to do with the value. When adding parallelism this can be especially helpful (no need to worry about another goroutine altering the value whilst you're using it). I use pointers for some things, esp more complex types with lots of methods, but for simple types like this which are basically just data I prefer the value-based approach.